### PR TITLE
openSSL compatibility 201711

### PR DESCRIPTION
**Pull request by kojo1**
Rebased, resolved conflict:
BN_mod_mul
SSL_CTX_clear_options
SSL_set_msg_callback_arg
X509_STORE_CTX_set_verify_cb
ERR_load_BIO_strings
SSL_CTX_set_alpn_protos
SSL_get_peer_cert_chain

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -26335,35 +26382,80 @@ WOLFSSL_API int wolfSSL_CTX_set1_curves_list(WOLFSSL_CTX* ctx, char* names)
...
+    if (ret == NULL) {
+        return NULL;
+    }
+    return XMEMCPY(ret, data, siz);
+}
+
+int wolfSSL_CTX_set_alpn_protos(WOLFSSL_CTX *ctx, const unsigned char *p,
+                            unsigned int p_len)
+{
+    if((void *)ctx->alpn_cli_protos != NULL)
```
Comment (cconlon) [inline src/ssl.c]: The WOLFSSL_CTX pointer should be checked for NULL before dereferencing to prevent a seg fault.

File: src/ssl.c [hunk 2]
```diff
@@ -26335,35 +26382,80 @@ WOLFSSL_API int wolfSSL_CTX_set1_curves_list(WOLFSSL_CTX* ctx, char* names)
...
 int wolfSSL_CTX_set_msg_callback(WOLFSSL_CTX *ctx, SSL_Msg_Cb cb)
 {
     WOLFSSL_STUB("SSL_CTX_set_msg_callback");
     (void)ctx;
     (void)cb;
     return WOLFSSL_FAILURE;
 }
+#endif
+
 int wolfSSL_set_msg_callback(WOLFSSL *ssl, SSL_Msg_Cb cb)
```
Comment (cconlon) [inline src/ssl.c]: Can you add a unit test for wolfSSL_set_msg_callback()?  Have you verified that this works as expected?

File: src/ssl.c [hunk 3]
```diff
@@ -11461,21 +11469,28 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
 
     long wolfSSL_CTX_get_options(WOLFSSL_CTX* ctx)
     {
-        (void)ctx;
         WOLFSSL_ENTER("wolfSSL_CTX_get_options");
         WOLFSSL_MSG("wolfSSL options are set through API calls and macros");
-
-        return 0;
+        return ctx->mask;
```
Comment (cconlon) [inline src/ssl.c]: We should check ctx for NULL first, to prevent possible seg fault.

File: src/ssl.c [hunk 4]
```diff
@@ -11461,21 +11469,28 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
...
+        if(ssl == NULL)return SSL_FAILURE;
+        ctx->mask = wolfSSL_set_options(ssl, opt);
+        wolfSSL_free(ssl);
+        return ctx->mask;
     }
 
+    long wolfSSL_CTX_clear_options(WOLFSSL_CTX* ctx, long opt)
+    {
+        WOLFSSL_ENTER("SSL_CTX_clear_options");
+        ctx->mask &= ~opt;
```
Comment (cconlon) [inline src/ssl.c]: We should check ctx against NULL before using.

File: src/ssl.c [hunk 5]
```diff
@@ -11564,6 +11579,12 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
             return ctx->error_depth;
         return WOLFSSL_FATAL_ERROR;
     }
+
+    void wolfSSL_X509_STORE_CTX_set_verify_cb(WOLFSSL_X509_STORE_CTX *ctx,
+                                  WOLFSSL_X509_STORE_CTX_verify_cb verify_cb)
+    {
+        ctx->verify_cb = verify_cb;
```
Comment (cconlon) [inline src/ssl.c]: We should check ctx against NULL first before using.

File: src/ssl.c [hunk 6]
```diff
@@ -17053,13 +17073,20 @@ unsigned long wolfSSL_set_options(WOLFSSL* ssl, unsigned long op)
...
 {
     WOLFSSL_ENTER("wolfSSL_get_options");
 
     return ssl->options.mask;
 }
 
+long wolfSSL_clear_options(WOLFSSL* ssl, long opt)
+{
+    WOLFSSL_ENTER("SSL_clear_options");
+    ssl->options.mask &= ~opt;
```
Comment (cconlon) [inline src/ssl.c]: We should check "ssl" against NULL before dereferencing.

File: tests/api.c [hunk 7]
```diff
@@ -10369,6 +10514,28 @@ static void test_wolfSSL_set_options(void)
     #endif /* defined(OPENSSL_EXTRA) && !defined(NO_CERTS) && \
              !defined(NO_FILESYSTEM) && !defined(NO_RSA) */
 }
+	  #if defined(OPENSSL_EXTRA) && !defined(NO_CERTS)
+static int verify_cb(int ok, X509_STORE_CTX *ctx)
+{
+	  (void) ok;
+		(void) ctx;
```
Comment (cconlon) [inline tests/api.c]: Can you adjust spacing here?  These few lines look to be inconsistent.

File: tests/api.c [hunk 8]
```diff
@@ -10369,6 +10514,28 @@ static void test_wolfSSL_set_options(void)
...
+}
+#endif
+
+static void test_wolfSSL_X509_STORE_CTX(void)
+{
+	  #if defined(OPENSSL_EXTRA) && !defined(NO_CERTS)
+	  X509_STORE_CTX *ctx   = NULL ;
+
+    printf(testingFmt, "test_wolfSSL_X509_STORE_CTX(()");
+		AssertNotNull(ctx = X509_STORE_CTX_new());
```
Comment (cconlon) [inline tests/api.c]: Can you fix spacing here?  Looks like there might be a few hard tabs.

File: src/ssl.c (hunk 3 above)
Comment (kojo1) [reply src/ssl.c]: Done.

File: src/ssl.c (hunk 4 above)
Comment (kojo1) [reply src/ssl.c]: Done

File: src/ssl.c (hunk 5 above)
Comment (kojo1) [reply src/ssl.c]: Done

File: src/ssl.c (hunk 6 above)
Comment (kojo1) [reply src/ssl.c]: Done

File: src/ssl.c (hunk 2 above)
Comment (kojo1) [reply src/ssl.c]: Yes, in test_wolfSSL_set_options, please find SSL_set_msg_callback.

File: src/ssl.c (hunk 1 above)
Comment (kojo1) [reply src/ssl.c]: Done

File: tests/api.c (hunk 7 above)
Comment (kojo1) [reply tests/api.c]: FIxed

File: tests/api.c (hunk 8 above)
Comment (kojo1) [reply tests/api.c]: FIxed

File: src/ssl.c [hunk 9]
```diff
@@ -11471,14 +11471,17 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
     {
         WOLFSSL_ENTER("wolfSSL_CTX_get_options");
         WOLFSSL_MSG("wolfSSL options are set through API calls and macros");
+        if(ctx == NULL)return SSL_FAILURE;
```
Comment (cconlon) [inline src/ssl.c]: To be consistent with our existing coding style, can you please place return calls on the following line, ie:

```
if (ctx == NULL)
    return SSL_FAILURE;
```

Can you also adjust this throughout this PR?  Thanks.

File: tests/api.c [hunk 10]
```diff
@@ -10335,6 +10355,95 @@ static void test_wolfSSL_BN(void)
...
 }
 
+#if defined(OPENSSL_EXTRA) && !defined(NO_CERTS) && \
+   !defined(NO_FILESYSTEM) && !defined(NO_RSA)
+#define TEST_ARG 0x1234
+static void msg_cb(int write_p, int version, int content_type,
+                   const void *buf, size_t len, SSL *ssl, void *arg)
+{
+    (void)write_p;
+		(void)version;
```
Comment (cconlon) [inline tests/api.c]: Please remove hard tabs in this function.

File: tests/api.c [hunk 11]
```diff
@@ -10343,11 +10452,40 @@ static void test_wolfSSL_set_options(void)
     SSL*     ssl;
     SSL_CTX* ctx;
 
+		unsigned char protos[] = {
+				7, 't', 'l', 's', '/', '1', '.', '2',
```
Comment (cconlon) [inline tests/api.c]: Please remove hard tabs in this function.

File: tests/api.c [hunk 12]
```diff
@@ -10361,6 +10499,13 @@ static void test_wolfSSL_set_options(void)
                                        SSL_OP_NO_TLSv1_2) == SSL_OP_NO_TLSv1_2);
     AssertTrue((SSL_set_options(ssl, SSL_OP_NO_COMPRESSION) &
                                SSL_OP_NO_COMPRESSION) == SSL_OP_NO_COMPRESSION);
+    AssertNull((SSL_clear_options(ssl, SSL_OP_NO_COMPRESSION) &
+                                       SSL_OP_NO_COMPRESSION));
+
+		AssertTrue(SSL_set_msg_callback(ssl, msg_cb) == SSL_SUCCESS);
```
Comment (cconlon) [inline tests/api.c]: Please remove hard tabs around this line.

File: tests/api.c [hunk 13]
```diff
@@ -10362,7 +10362,7 @@ static void msg_cb(int write_p, int version, int content_type,
                    const void *buf, size_t len, SSL *ssl, void *arg)
 {
     (void)write_p;
-		(void)version;
+    (void)version;
 		(void)content_type;
```
Comment (cconlon) [inline tests/api.c]: Looks like there are a few more hard tabs.  There are a few helpful ways that I have found to identify hard tabs:

1)  In the GitHub web view of a PR, view the "Files changed".  This allows you to see spacing differences.

2)  In vim, you can search for all hard tabs using:

/\t

This is referenced here:
http://vim.wikia.com/wiki/Highlight_unwanted_spaces

**Comment by kojo1** [comment]
I hope this version looks fine. Thank you for your review.

File: src/ssl.c (hunk 4 above)
Comment (cconlon) [reply src/ssl.c]: I don't see that a CTX NULL check was added in this function (wolfSSL_CTX_clear_options()).

File: src/ssl.c (hunk 6 above)
Comment (cconlon) [reply src/ssl.c]: I don't see where a ssl NULL check was added in this function (wolfSSL_clear_options()).

**Comment by cconlon** [comment]
Hi Kojo,

Just a reminder to address the last two comments regarding wolfSSL_CTX_clear_options() and wolfSSL_clear_options() argument sanitization.

Thanks,
Chris

File: src/ssl.c (hunk 4 above)
Comment (cconlon) [reply src/ssl.c]: Please add an argument check for "ctx" against NULL here, thanks.

**Comment by cconlon** [comment]
Thanks for adding a check for wolfSSL_clear_options().  Can you also add one for wolfSSL_CTX_clear_options()?

**Comment by cconlon** [comment]
Thanks!  I'm going to wait to merge this until after the release, but looks good.

**Pull request merged**